### PR TITLE
feat: `_makePayments` primitive testing (SC-3754) 

### PR DESCRIPTION
**Pull request by JGcarv**
# Description

# Integrations Checklist

- [ ] Have any function signatures changed? If yes, outline below.
- [ ] Have any features changed or been added? If yes, outline below.
- [ ] Have any events changed or been added? If yes, outline below.
- [ ] Has all documentation been updated?

# Changelog

### CODE REVIEW

File: contracts/test/LoanPrimitive.t.sol [hunk 1]
```diff
@@ -1008,3 +1008,112 @@ contract LoanPrimitiveRepossessTest is DSTest {
...
+
+contract LoanPrimitiveMakePaymentTest is DSTest {
+
+    Hevm hevm;
+
+    LoanPrimitiveHarness loan;
+    MockERC20            collateralAsset;
+    MockERC20            fundsAsset;
+
+    uint256 constant MAX_TIME = 10 * 365 days; // Assumed reasonable upper limit for payment intervals and grace periods
```
Comment (deluca-mike) [inline contracts/test/LoanPrimitive.t.sol:1020]:
```suggestion
    uint256 constant MAX_TIME = 10 * 365 days;  // Assumed reasonable upper limit for payment intervals and grace periods
```

File: contracts/test/LoanPrimitive.t.sol [hunk 2]
```diff
@@ -1008,3 +1008,112 @@ contract LoanPrimitiveRepossessTest is DSTest {
...
+        uint256 interestRate_, 
+        uint256 lateFeeRate_, 
+        uint256 paymentInterval_, 
+        uint256 paymentsRemaining_,
+        uint256 numberOfPayments_
+        ) 
+        external 
+    {
+
+        uint256 endingPrincipal = _constrictToRange(endingPrincipal_, 0, PRINCIPAL_REQUESTED);
```
Comment (deluca-mike) [inline contracts/test/LoanPrimitive.t.sol]: Can reuse stack item `endingPrincipal_` as `endingPrincipal_` is no longer used, and just makes sense as to reduce chances of `STACK_TOO_DEEP`.
```suggestion
    {
        uint256 endingPrincipal_ = _constrictToRange(endingPrincipal_, 0, PRINCIPAL_REQUESTED); 
```

File: contracts/test/LoanPrimitive.t.sol [hunk 3]
```diff
@@ -1008,3 +1008,112 @@ contract LoanPrimitiveRepossessTest is DSTest {
...
+        ) 
+        external 
+    {
+
+        uint256 endingPrincipal = _constrictToRange(endingPrincipal_, 0, PRINCIPAL_REQUESTED); 
+        uint256 interestRate = _constrictToRange(interestRate_, 0, 10_000); 
+        uint256 lateFeeRate = _constrictToRange(lateFeeRate_, 0, 10_000); 
+        uint256 paymentInterval = _constrictToRange(paymentInterval_, 100, MAX_TIME); 
+        uint256 paymentsRemaining = _constrictToRange(paymentsRemaining_, 1, 120);
+        uint256 numberOfPayments = _constrictToRange(numberOfPayments_, 1, paymentsRemaining);
```
Comment (deluca-mike) [inline contracts/test/LoanPrimitive.t.sol]: Same for all of these. Just overwrite the input.

File: contracts/test/LoanPrimitive.t.sol (hunk 1 above)
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:1020]: Should use `10_000 * 365 days` as upper bound. 10 days is too small

File: contracts/test/LoanPrimitive.t.sol [hunk 4]
```diff
@@ -1008,3 +1008,112 @@ contract LoanPrimitiveRepossessTest is DSTest {
...
+
+        loan.drawdownFunds(PRINCIPAL_REQUESTED, address(this));
+
+        /********************/
+        /*** Make Payment ***/
+        /********************/
+
+        assertEq(loan.drawableFunds(), 0);
+        assertEq(loan.claimableFunds(), 0);
+        assertEq(loan.principal(), PRINCIPAL_REQUESTED);
```
Comment (deluca-mike) [inline contracts/test/LoanPrimitive.t.sol]:
```suggestion
        assertEq(loan.drawableFunds(),  0);
        assertEq(loan.claimableFunds(), 0);
        assertEq(loan.principal(),      PRINCIPAL_REQUESTED);
```

File: contracts/test/LoanPrimitive.t.sol [hunk 5]
```diff
@@ -1008,3 +1008,112 @@ contract LoanPrimitiveRepossessTest is DSTest {
...
+
+    Hevm hevm;
+
+    LoanPrimitiveHarness loan;
+    MockERC20            collateralAsset;
+    MockERC20            fundsAsset;
+
+    uint256 constant MAX_TIME = 10 * 365 days; // Assumed reasonable upper limit for payment intervals and grace periods
+    uint256 constant COLLATERAL_REQUIRED = 300_000;
+    uint256 constant PRINCIPAL_REQUESTED = 1_000_000;
```
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:1022]: Should use one
```js
uint256 constant MAX_TOKEN_AMOUNT = 1e12 * 10 ** 18;  // 1 trillion of a token with 18 decimals (assumed reasonable upper limit for token amounts)
``` 
value

File: contracts/test/LoanPrimitive.t.sol [hunk 6]
```diff
@@ -1008,3 +1008,112 @@ contract LoanPrimitiveRepossessTest is DSTest {
...
+        return min_ == max_ ? max_ : input_ % (max_ - min_) + min_;
+    }
+
+    function test_makePayment(
+        uint256 endingPrincipal_, 
+        uint256 interestRate_, 
+        uint256 lateFeeRate_, 
+        uint256 paymentInterval_, 
+        uint256 paymentsRemaining_,
+        uint256 numberOfPayments_
```
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:1042]: Need to fuzz `collateralRequired_` and `principalRequested_`

Comment (lucas-manuel) [reply contracts/test/LoanPrimitive.t.sol:1042]: And constrict them to `0, MAX_TOKEN_AMOUNT`

File: contracts/test/LoanPrimitive.t.sol [hunk 7]
```diff
@@ -1008,3 +1008,112 @@ contract LoanPrimitiveRepossessTest is DSTest {
...
+    ) 
+        external 
+    {
+
+        uint256 endingPrincipal   = _constrictToRange(endingPrincipal_,   0,   PRINCIPAL_REQUESTED);
```
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:1047]: `0, principalRequested`

File: contracts/test/LoanPrimitive.t.sol [hunk 8]
```diff
@@ -1008,3 +1008,112 @@ contract LoanPrimitiveRepossessTest is DSTest {
...
+        fundsAsset.mint(address(loan), totalPayment);
+
+        loan.makePayments(numberOfPayments);
+
+        assertEq(loan.drawableFunds(),      0);      
+        assertEq(loan.claimableFunds(),     totalPayment);     
+        assertEq(loan.collateral(),         COLLATERAL_REQUIRED);         
+        assertEq(loan.nextPaymentDueDate(), start + loan.paymentInterval() * 2);  // Made a payment, so paymentInterval moved
+        assertEq(loan.principal(),          PRINCIPAL_REQUESTED - principal);          
+        assertEq(loan.paymentsRemaining(),  paymentsRemaining - numberOfPayments);
```
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:1116]: UPdate assertions accordingly

File: contracts/test/LoanPrimitive.t.sol [hunk 9]
```diff
@@ -1008,3 +1008,112 @@ contract LoanPrimitiveRepossessTest is DSTest {
...
+
+        uint256 totalPayment = principal + interest + lateFees;
+
+        fundsAsset.mint(address(loan), totalPayment);
+
+        loan.makePayments(numberOfPayments);
+
+        assertEq(loan.drawableFunds(),      0);      
+        assertEq(loan.claimableFunds(),     totalPayment);     
+        assertEq(loan.collateral(),         COLLATERAL_REQUIRED);
```
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:1113]: Don't need collateral assertion

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: contracts/test/LoanPrimitive.t.sol [hunk 10]
```diff
@@ -1170,3 +1171,147 @@ contract LoanPrimitiveClaimFundsTest is DSTest {
...
+        uint256 paymentsRemaining_,
+        uint256 numberOfPayments_
+    ) 
+        external 
+    {
+        principalRequested_ = _constrictToRange(principalRequested_, 1,   MAX_TOKEN_AMOUNT);
+        collateralRequired_ = _constrictToRange(collateralRequired_, 1,   MAX_TOKEN_AMOUNT);
+        endingPrincipal_    = _constrictToRange(endingPrincipal_,    0,   principalRequested_); 
+        interestRate_       = _constrictToRange(interestRate_,       0,   10_000); 
+        lateFeeRate_        = _constrictToRange(lateFeeRate_,        0,   10_000);
```
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:1231]: 10_000 * 100

File: contracts/test/LoanPrimitive.t.sol [hunk 11]
```diff
@@ -497,7 +497,7 @@ contract LendPrimitivePostAndRemoveCollateralTest is DSTest {
         loan            = new LoanPrimitiveHarness();
     }
 
-    function _initializeLoanWithCollateralRequested(uint256 collateralRequested_) internal {
```
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:500]: `_initializeLoanWithCollateralRequired`

File: contracts/test/LoanPrimitive.t.sol [hunk 12]
```diff
@@ -1170,3 +1171,147 @@ contract LoanPrimitiveClaimFundsTest is DSTest {
...
+        collateralAsset = new MockERC20("CollateralAsset", "CA", 0);
+        fundsAsset      = new MockERC20("FundsAsset",      "FA", 0);
+        loan            = new LoanPrimitiveHarness();
+    }
+
+    function _constrictToRange(uint256 input_, uint256 min_, uint256 max_) internal pure returns (uint256 output_) {
+        return min_ == max_ ? max_ : input_ % (max_ - min_) + min_;
+    }
+
+    function _initializeLoanWithpaymentsRemaining(uint256 paymentsRemaining_) internal {
```
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:1198]: WithPayments

File: contracts/test/LoanPrimitive.t.sol [hunk 13]
```diff
@@ -1170,3 +1171,147 @@ contract LoanPrimitiveClaimFundsTest is DSTest {
...
+
+        /********************/
+        /*** Make Payment ***/
+        /********************/
+
+        assertEq(loan.drawableFunds(),  0);
+        assertEq(loan.claimableFunds(), 0);
+        assertEq(loan.principal(),      principalRequested_);
+
+        hevm.warp(start + paymentInterval_ * numberOfPayments_);
```
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:1275]: We don't need this warp

File: contracts/test/LoanPrimitive.t.sol [hunk 14]
```diff
@@ -1170,3 +1171,147 @@ contract LoanPrimitiveClaimFundsTest is DSTest {
...
+            uint256(365 days / 6),
+            paymentsRemaining_
+        ];
+
+        uint256[2] memory requests = [uint256(300_000), uint256(1_000_000)];
+
+        loan.initialize(address(1), assets, parameters, requests);
+    }
+
+    function test_makePayment(
```
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:1215]: makePayments

File: contracts/test/LoanPrimitive.t.sol [hunk 15]
```diff
@@ -1170,3 +1171,147 @@ contract LoanPrimitiveClaimFundsTest is DSTest {
...
+        assertEq(loan.claimableFunds(),     totalPayment);             
+        assertEq(loan.nextPaymentDueDate(), start + paymentInterval_ * (numberOfPayments_ + 1));  // Made a payment, so paymentInterval moved
+        assertEq(loan.principal(),          principalRequested_ - principal);          
+        assertEq(loan.paymentsRemaining(),  paymentsRemaining_ - numberOfPayments_);  
+    }
+
+    function test_makePayment_morePaymentsThanRemaining(uint256 paymentsRemaining_) external {
+        paymentsRemaining_ = _constrictToRange(paymentsRemaining_, 1, 120);
+
+        //Only 1 payment to do
```
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:1305]: // Only

Comment (lucas-manuel) [reply contracts/test/LoanPrimitive.t.sol:1305]: Actually not sure what this comment means

File: contracts/test/LoanPrimitive.t.sol [hunk 16]
```diff
@@ -1170,3 +1171,147 @@ contract LoanPrimitiveClaimFundsTest is DSTest {
...
+        _initializeLoanWithpaymentsRemaining(paymentsRemaining_);
+
+        uint256 start = block.timestamp;
+        hevm.warp(loan.nextPaymentDueDate());
+
+        //Provide more than enough tokens
```
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:1311]: // Provide

File: contracts/test/LoanPrimitive.t.sol [hunk 17]
```diff
@@ -1170,3 +1171,147 @@ contract LoanPrimitiveClaimFundsTest is DSTest {
...
+        assertEq(loan.paymentsRemaining(),  paymentsRemaining_ - numberOfPayments_);  
+    }
+
+    function test_makePayment_morePaymentsThanRemaining(uint256 paymentsRemaining_) external {
+        paymentsRemaining_ = _constrictToRange(paymentsRemaining_, 1, 120);
+
+        //Only 1 payment to do
+        _initializeLoanWithpaymentsRemaining(paymentsRemaining_);
+
+        uint256 start = block.timestamp;
```
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:1308]: We don't need start

File: contracts/test/LoanPrimitive.t.sol [hunk 18]
```diff
@@ -1170,3 +1171,147 @@ contract LoanPrimitiveClaimFundsTest is DSTest {
...
+        hevm.warp(loan.nextPaymentDueDate());
+
+        //Provide more than enough tokens
+        fundsAsset.mint(address(loan), MAX_TOKEN_AMOUNT);
+
+        try loan.makePayments(paymentsRemaining_ + 1) { assertTrue(false);} catch {assertTrue(true);}
```
Comment (lucas-manuel) [inline contracts/test/LoanPrimitive.t.sol:1314]: Spaces on either side of {}: `{ assertTrue(false); }`, empty block for `catch`
```js
try loan.makePayments(paymentsRemaining_ + 1) { assertTrue(false); } catch {}
```

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: contracts/LoanPrimitive.sol [hunk 19]
```diff
@@ -122,7 +122,7 @@ contract LoanPrimitive {
         _drawableFunds = _drawableFunds + _getUnaccountedAmount(_fundsAsset) - totalAmountPaid;
 
         _claimableFunds     += totalAmountPaid;
-        _nextPaymentDueDate += _paymentInterval;
+        _nextPaymentDueDate += _paymentInterval * numberOfPayments_;
```
Comment (deluca-mike) [inline contracts/LoanPrimitive.sol:125]: 💯

**Review by deluca-mike — APPROVED** [review]

**Pull request merged**